dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -144,7 +160,49 @@ private OptionalInt incrementHostIfNecessary(int currentIndex) {
         return saved ? OptionalInt.of(nextIndex) : OptionalInt.empty(); // we've moved on already
     }
 
-    interface NodeList {
+    Optional<PinUntilErrorChannel> reloadableFrom(Object other) {
```

Comment:
The object casting logic is a bit difficult to follow, perhaps we can provide a `NodeSelector` interface with the functionality we need, and implement it for round-robin as well? Round robin could effectively do what we're already doing, but using a volatile delegate instance that's recreated when nodes change.

Reply:
I think the "reloadableFrom" then "liveReloadNewInstance" dance is really a key bit of the PR, so would like to try and make it as clear as possible. The reason I've got the instanceof check is because in the case where there's only a single uri, we have the clever short-circuiting thing:

```
            if (channels.size() == 1) {
                return channels.get(0); // no fancy node selection heuristic can save us if our one node goes down
            }
```

Also I want to make it clear that if someone was on ROUND_ROBIN, then there's no way to live-reload our way to PIN_UNTIL_ERROR, we just have to recreated.  That's where the `reloadableFrom` method would just throw up its hands and say I can't work worth with this, just use a fresh instance.